lib/service/version.go

```diff
@@ -33,63 +31,47 @@ const (
 	majorVersionConstraint = 18
 )
 
-var (
-	// errMajorVersionUpgrade default error for restricting upgrade major version.
-	errMajorVersionUpgrade = errors.New(`Teleport supports upgrade only one major version behind. ` +
-		`See: https://goteleport.com/docs/upgrading/overview/#component-compatibility`)
-)
-
 // validateAndUpdateTeleportVersion validates that the major version persistent in the backend
 // meets our upgrade compatibility guide.
 func (process *TeleportProcess) validateAndUpdateTeleportVersion(currentVersion string) error {
-	currentMajor, err := getMajorVersion(currentVersion)
+	currentMajor, err := utils.MajorVersion(currentVersion)
 	if err != nil {
 		return trace.Wrap(err)
 	}
 
-	st, err := process.storage.GetState(process.GracefulExitContext(), types.RoleInstance)
+	localVersion, err := process.storage.GetTeleportVersion(process.GracefulExitContext())
 	if trace.IsNotFound(err) {
-		// If the Instance state has not been found, it means this is the first launch,
-		// and we don't need to check the version since this is the initial launch.
-		return nil
-	} else if err != nil {
-		return trace.Wrap(err)
-	}
-
-	if st.Spec.LocalVersion == "" {
-		if currentMajor >= majorVersionConstraint {
-			return trace.Wrap(errMajorVersionUpgrade, "Teleport is going to update to version %s, but before must be upgraded to %d.x.x first",
+		// We have to ensure that the process database is newly created before
+		//applying the majorVersionConstraint check.
+		_, err := process.storage.GetState(process.GracefulExitContext(), types.RoleAdmin)
+		if !trace.IsNotFound(err) && err != nil {
+			return trace.Wrap(err)
+		}
+		if currentMajor >= majorVersionConstraint && !trace.IsNotFound(err) {
+			return trace.BadParameter("Unsupported upgrade path detected: to %v. "+
+				"Teleport supports direct upgrades to the next major version only.\n Please upgrade "+
+				"your cluster to version %d.x.x first. See compatibility guarantees for details: "+
+				"https://goteleport.com/docs/upgrading/overview/#component-compatibility.",
```

Comment:
I think this error message is incorrect.

If you were on Teleport 15 and you tried to upgrade straight to Teleport 18, wouldn't it say "Please upgrade your cluster to version 17.x.x first" ?? 

If so, that's not right, we should be telling them to upgrade to v16 first.

Reply:
in this specific case we couldn't know exact version of previous setup, I would probably change the message with example from docs instead of this version suggestion